`NormalizeLets` -- inlining at inner nesting levels

ggevay opened the pull request:
### Motivation

  * This PR fixes a recognized bug: https://github.com/MaterializeInc/materialize/issues/18889

### Tips for reviewer

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

aalexandrov commented on src/transform/src/normalize_lets.rs:
```diff
@@ -438,6 +438,28 @@ mod inlining {
+        let mut worklist = vec![&mut *expr];
+        while let Some(expr) = worklist.pop() {
+            inline_lets_core(expr, inline_mfp)?;
+            // We descend only into `LetRec` nodes, because they have been already normalized to
```
nit: I think the comment is can be slightly improved:

> We descend into top-level `LetRec` nodes (which will exist if inline_lets_core has done some work)
> in order to normalize recursive `LetRec` nodes.

ggevay replied:
> which will exist if inline_lets_core has done some work

Actually, there might be a `LetRec` node here even if `inline_lets_core` didn't do any work.

I edited the comment a bit. So, why it is enough to descend only into `LetRec` nodes, is because `promote_let_rec` ensured that if there are any `LetRec` nodes in a subtree, then one of those has to be at the root. In other words, if there are multiple `LetRec` nodes in a tree, then you can get to all of them by just descending into `LetRec` nodes. 

This is the new comment:
```
            // We descend only into `LetRec` nodes, because `promote_let_rec` ensured that all
            // `LetRec` nodes are clustered near the root. This means that we can get to all the
            // `LetRec` nodes by just descending into `LetRec` nodes, as there can't be any other
            // nodes between them.
```

aalexandrov commented on test/sqllogictest/transform/normalize_lets.slt:
```diff
@@ -63,6 +63,45 @@ Explained Query:
+## Test inlining at an inner nesting level. (#18889)
+## `bar` is used only in `baz`, so it should be inlined. The inner WMR should have only one cte.
+query T multiline
+EXPLAIN WITH MUTUALLY RECURSIVE
```
Do we also currently cover the case where we have a non-recurisve (`Let`) CTE at the root or non-recursive CTEs between the recrusive (`LetRec`) CTEs?

```sql
WITH 
  cte0 AS (
    -- non-recursive CTE
  )
  ( 
    WITH MUTUALLY RECURSIVE
    -- possibly with further non-recursive CTEs
  )

ggevay replied:
Yes, we cover the case where the situation that you described occurs just before the entire `NormalizeLets`. I'll add a test for this.

But note that at the point where `inline_lets` is called, we don't have any `Let` nodes anywhere in the tree. (I might add an assert for this.) At this point, all of them have been normalized to be `LetRec` nodes. (`promote_let_rec` does this through `digest_lets_helper`.) And then if there is no actual recursion, then they get turned back into a stack of `Let`s at the end of `action` at the `// Disassemble `LetRec` into a `Let` stack if possible.`

ggevay commented:
Thanks for the review!
- I tried to improve the comment a bit. @aalexandrov, can you please check whether it's better now?
- I added an assert for no `Let` directly after `promote_let_rec`.
- I added a test for having a `Let` before a `LetRec`, where the cte from the `Let` has to be inlined into the `LetRec`.
